doc/arch/020-questionnaires-aka-dynamic-forms.md

```diff
@@ -0,0 +1,127 @@
+1. Mobile has a questionnaire of given type.
+1. Mobile has questionnaire responses for given type.
+1. Config of a questionnaire is enabled for User's facility, which mobile derives from [the facility sync API](https://api.simple.org/api-docs/index.html#tag/facility).
+
+### 4. Displaying a list of forms
+![questionnaire-responses-list](resources/questionnaire-responses-list.png)
+
+When user clicks on a form type from home page, Mobile queries its table for responses of given questionnaire type. Mobile has some custom logic on ordering and style of the forms:
+1. Mobile orders forms in descending order of months by checking for `content["month_date"]` key.
+1. Mobile deduces submission status by checking for `content["submitted"]` key.
+1. Mobile localizes month_date before displaying them.
+1. This implies that `month_date` & `submitted` keys are _"mandatory/static"_ in monthly questionnaire responses.
+1. Mobile has display & translation logic for these mandatory fields.
+
+### 5. User clicks on a form
+![questionnaire-view](resources/questionnaire-view.png)
+
+When user clicks on a form, Mobile renders UI of a questionnaire based on the layout. Current version of DSL has provisions for `integer, string, date & radio-button` input types, and `header/sub-header/paragraph/checklist` display types; and can be extended to more types in future.
+
+### 6. User submits a form
+When user submits a form, Mobile updates the `content`, `questionnaire_id`(to record the layout which was displayed to a user), `last_updated_by_user_id` & `facility_id` for a questionnaire response.
+
+A questionnaire's layout contains a `link_id` for input fields. Mobile uses them to display values in UI and record responses in the content key. 
+A specimen content key's value looks like:
+```json
+{
+  "submitted": true,
+  "month_date": "08-2023",
+  "monthly_screening_report.total_bp_checks.male": 10,
+  "monthly_screening_report.comments": "All good"
+}
+```
+
+### 7. Handling concurrent updates
+When server receives different questionnaire_responses for the same ID, it merges the 2 records in the following way:
+1. Latest `device_updated_at` takes precedence for every field, except `content`.
+1. A UNION of existing & new `content` is taken.
+1. For conflicting keys within content, latest update takes precedence.
+1. Consider this specimen workflow:
+    1. User #1 fills keys A, B, C & D on 6 Dec
+    1. User #2 fills keys C, D, E & F on 7 Dec
+    1. Post union, DB will contain keys A, B, C, D, E & F
+    1. C, D will be User #2’s data.
+
+### 8. User changes locale
+
+For most screens, mobile stores the translations for all languages. Since a questionnaire's layout is driven by server, translations are also stored on server-side. When user changes a locale, a sync gets triggerd on mobile. We leveraged this behaviour of mobile to communicate layouts with new translations from server to mobile. 
```

Comment:
typo fix and prose suggestion:
```suggestion
For most screens, mobile stores the translations for all languages. Since a questionnaire's layout is driven by server, translations are also stored on server-side. When user changes a locale, a sync gets triggered on mobile. We leverage this behaviour of mobile to communicate layouts with new translations from server to mobile. 
```